docs/architecture/adr-071-bank-v2.md

```diff
@@ -0,0 +1,183 @@
+    OnBalanceOf func(ctx context.Context, addr []byte, denom string) (sdk.Coins, error)
+}
+
+func (k *BankKeeper) MintCoins(ctx context.Context, moduleName string, amt sdk.Coins) error {
+    if k.OnMint != nil {
+        if err := k.OnMint(ctx, moduleName, amt); err != nil {
+            return err
+        }
+    }
+}
+
+func (k *BankKeeper) BurnCoins(ctx context.Context, moduleName string, amt sdk.Coins) error {
+    if k.OnBurn != nil {
+        if err := k.OnBurn(ctx, moduleName, amt); err != nil {
+            return err
+        }
+    }
+}
+
+func (k *BankKeeper) SendCoins(ctx context.Context, fromAddr, toAddr []byte, amt sdk.Coins) error {
+    if k.OnTransfer != nil {
+        if err := k.OnTransfer(ctx, fromAddr, toAddr, amt); err != nil {
+            return err
+        }
+    }
+}
+
+func (k *BankKeeper) GetBalance(ctx context.Context, addr []byte, denom string) (sdk.Coins, error) {
+    if k.OnBalanceOf != nil {
+        coins, err := k.OnBalanceOf(ctx, addr, denom);
+        if err != nil {
+            return nil, err
+        }
+
+        return coins, err
+    }
+}
+```
+
+Messages for Admin Operations:
+
+To facilitate admin operations like `Mint`, `Burn`, and `Transfer`, we can define specific messages that can be invoked by the admin account of each denom. These operations will be internal and not exposed to end clients directly.
+
+```protobuf
+message Mint {
+  string denom = 1;
+  string to = 2;
+  string amount = 3;
+}
+
+message Burn {
+  string denom = 1;
+  string from = 2;
+  string amount = 3;
+}
+
+message Transfer {
+  string denom = 1;
+  string from = 2;
+  string to = 3;
+  string amount = 4;
+}
+```
+
+If admin account defines these callbacks, `x/bank` will call `x/accounts` `MsgExecute` with the `OnTransfer`, `GetBalance` or `GetSupply` callbacks:
+
+```protobuf 
+// can error if transfer isn't allowed
+message OnTransfer {
+  string denom = 1;
+  string from = 2;
+  string to = 3;
+  string amount = 4;
+}
+
+// returns GetBalanceResponse
+// can override state management in x/bank
+// otherwise x/bank manages balance state
+// if GetBalance is defined, GetSupply also needs to be defined
+
+message GetBalance {
+  string denom = 1;
+  string account = 2;
+}
+
+message GetBalanceResponse {
+  string amount = 1;
+}
+
+message GetSupply {
+  string denom = 1;
+}
+
+message GetSupplyResponse {
+  string amount = 1;
+}
+```
+
+x/bank is responsible for emitting send, mint and burn events for consistency.
+
+## Migration Plans
+
+Bank is a widely used module, so getting a v2 needs to be thought thoroughly. In order to not force all dependencies to immediately migrate to bank/v2, the same _upgrading_ path will be taken as for the `gov` module.
+
+This means `cosmossdk.io/bank` will stay one module and there won't be a new `cosmossdk.io/bank/v2` go module. Instead the bank protos will be versioned from `v1beta1` (current bank) to `v2`.
```

Comment:
i dont think we should do this since we dont want to support all the queries there right now? should we assume we have an offchain indexer in this sense and make sure we dont store extra data for queries?